core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -695,11 +690,16 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       assert (vlen % 8 == 0);
       assert (longArray != null);
 
-      if (numKeys == MAX_CAPACITY
-        // The map could be reused from last spill (because of no enough memory to grow),
-        // then we don't try to grow again if hit the `growthThreshold`.
-        || !canGrowArray && numKeys > growthThreshold) {
-        return false;
+      if (numKeys >= growthThreshold) {
+        if (longArray.size() / 2 == MAX_CAPACITY) {
```

Comment:
This does not look correct as per documentation of MAX_CAPACITY.
Actual number of keys == MAX_CAPACITY (so that total number of entries in longArray is MAX_CAPACITY * 2)